core/services/workflows/syncer/workflow_registry.go

```diff
@@ -2,39 +2,366 @@ package syncer
+	QueryKey(
+		context.Context,
+		types.BoundContract,
+		query.KeyFilter,
+		query.LimitAndSort,
+		any,
+	) ([]types.Sequence, error)
+}
+
+// ContractEventPollerConfig is the configuration needed to poll for events on a contract.  Currently
+// requires the ContractEventName.
+//
+// TODO(mstreet3): Can we query all contract events at once and tag each with an event type?
+//
+// TODO(mstreet3): Use LookbackBlocks instead of StartBlockNum
+type ContractEventPollerConfig struct {
+	ContractName      string
+	ContractAddress   string
+	ContractEventName string
+	StartBlockNum     uint64
+	QueryCount        uint64
+}
+
+// FetcherFunc is an abstraction for fetching the contents stored at a URL.
+type FetcherFunc func(ctx context.Context, url string) ([]byte, error)
+
+// WorkflowRegistrySyncer is the public interface of the package.
+type WorkflowRegistrySyncer interface {
+	services.Service
+}
+
+var _ WorkflowRegistrySyncer = (*workflowRegistry)(nil)
+
+// workflowRegistry is the implementation of the WorkflowRegistrySyncer interface.
+type workflowRegistry struct {
 	services.StateMachine
+	stopCh   services.StopChan
+	lggr     logger.Logger
+	orm      WorkflowRegistryDS
+	reader   ContractReader
+	gateway  FetcherFunc
+	wg       sync.WaitGroup
+	ticker   <-chan struct{}
+	eventsCh chan WorkflowRegistryEventResponse
+	handler  Handler
+	cfg      ContractEventPollerConfig
 }
 
-func (w *WorkflowRegistry) Start(ctx context.Context) error {
-	return nil
+// WithTicker allows external callers to provide a ticker to the workflowRegistry.  This is useful
+// for overriding the default tick interval.
+func WithTicker(ticker <-chan struct{}) func(*workflowRegistry) {
+	return func(wr *workflowRegistry) {
+		wr.ticker = ticker
+	}
 }
 
-func (w *WorkflowRegistry) Close() error {
-	return nil
+// NewWorkflowRegistry returns a new workflowRegistry.
+// Only queries for WorkflowRegistryForceUpdateSecretsRequestedV1 events.
+func NewWorkflowRegistry(
+	lggr logger.Logger,
+	orm WorkflowRegistryDS,
+	reader ContractReader,
+	gateway FetcherFunc,
+	addr string,
+	opts ...func(*workflowRegistry),
+) *workflowRegistry {
+	wr := &workflowRegistry{
+		lggr:    lggr.Named(name),
+		orm:     orm,
+		reader:  reader,
+		gateway: gateway,
+		cfg: ContractEventPollerConfig{
+			ContractName:      ContractName,
+			ContractAddress:   addr,
+			ContractEventName: ContractEventName,
+			QueryCount:        20,
+			StartBlockNum:     0,
+		},
+		stopCh:   make(services.StopChan),
+		eventsCh: make(chan WorkflowRegistryEventResponse),
+	}
+	wr.handler = newEventHandler(wr.lggr, wr.orm, wr.gateway)
+	for _, opt := range opts {
+		opt(wr)
+	}
+	return wr
+}
+
+// Start starts the workflowRegistry.  It starts two goroutines, one for querying the contract
+// and one for handling the events.
+func (w *workflowRegistry) Start(_ context.Context) error {
+	return w.StartOnce(w.Name(), func() error {
+		ctx, _ := w.stopCh.NewCtx()
```

Comment:
You can't ignore the cancel here; that's a goroutine that will leak.

You can avoid this by creating separate contexts inside each goroutine, eg.

```
ctx, cancel := w.stopCh.NewCtx()
defer cancel()
defer w.wg.Done()
w.queryLoop(ctx, w.cfg)
```